nodeup/pkg/model/hooks.go

```diff
@@ -17,53 +17,130 @@ limitations under the License.
+func (h *HookBuilder) Build(c *fi.ModelBuilderContext) error {
+	// we keep a list of hooks name so we can allow local instanceGroup hooks override the cluster ones
+	hookNames := make(map[string]bool, 0)
+	for i, spec := range []*[]kops.HookSpec{&h.InstanceGroup.Spec.Hooks, &h.Cluster.Spec.Hooks} {
+		for _, hook := range *spec {
+			isInstanceGroup := i == 0
+			// filter out on master and node flags if required
+			if (hook.MasterOnly && !h.IsMaster) || (hook.NodeOnly && h.IsMaster) {
+				continue
+			}
+			// i dont want to effect those whom are already using the hooks, so i'm gonna try an keep the name for now
+			// i.e. use the default naming convention - kops-hook-<index>, only those using the Name or hooks in IG should alter
+			var name string
+			switch hook.Name {
+			case "":
+				name = fmt.Sprintf("kops-hook-%d", i)
+				if isInstanceGroup {
+					name = fmt.Sprintf("%s-ig", name)
+				}
+			default:
+				name = hook.Name
+			}
+
+			if _, found := hookNames[name]; found {
+				glog.V(2).Infof("Skipping the hook: %v as we've already processed a similar service name", name)
+				continue
+			}
+			hookNames[name] = true
+
+			// are we disabling the service?
+			if hook.Disabled {
+				enabled := false
+				managed := true
+				c.AddTask(&nodetasks.Service{
+					Name:        hook.Name,
+					ManageState: &managed,
+					Enabled:     &enabled,
+					Running:     &enabled,
+				})
+				continue
+			}
+
+			service, err := h.buildSystemdService(name, &hook)
+			if err != nil {
+				return err
+			}
+
+			if service != nil {
+				c.AddTask(service)
+			}
 		}
-		c.AddTask(service)
 	}
 
 	return nil
 }
 
-func (b *HookBuilder) buildSystemdService(name string, hook *kops.HookSpec) (*nodetasks.Service, error) {
-	// We could give the container a kubeconfig, but we would probably do better to have a real pod / daemonset / job at that point
-	execContainer := hook.ExecContainer
-	if execContainer == nil {
-		glog.Warningf("No ExecContainer found for hook: %v", hook)
+// buildSystemdService is responsible for generating the service
+func (h *HookBuilder) buildSystemdService(name string, hook *kops.HookSpec) (*nodetasks.Service, error) {
+	// perform some basic validation
+	if hook.ExecContainer == nil && hook.Manifest == "" {
+		glog.Warningf("hook: %s has neither a raw unit or exec image configured", name)
 		return nil, nil
 	}
+	if hook.ExecContainer != nil {
+		if err := isValidExecContainerAction(hook.ExecContainer); err != nil {
+			glog.Warningf("invalid hook action, name: %s, error: %v", name, err)
+			return nil, nil
+		}
+	}
+	// build the base unit file
+	unit := &systemd.Manifest{}
+	unit.Set("Unit", "Description", "Kops Hook "+name)
 
-	image := strings.TrimSpace(execContainer.Image)
-	if image == "" {
-		glog.Warningf("No Image found for hook: %v", hook)
-		return nil, nil
+	// add any service dependencies to the unit
+	for _, x := range hook.Requires {
+		unit.Set("Unit", "Requires", x)
```

Comment:
We should definitely set `requires docker` automatically - are there any others you need?